internal/resolution/client/npm_registry_client.go

```diff
@@ -0,0 +1,285 @@
+		2*len(dependencies.BundleDependencies)
+	deps := make([]resolve.RequirementVersion, 0, depCount)
+	addDeps := func(ds map[string]string, t dep.Type) {
+		for name, req := range ds {
+			typ := t.Clone()
+			if r, ok := strings.CutPrefix(req, "npm:"); ok {
+				// This dependency is aliased, add it as a
+				// dependency on the actual name, with the
+				// KnownAs attribute set to the alias.
+				typ.AddAttr(dep.KnownAs, name)
+				name = r
+				req = ""
+				if i := strings.LastIndex(r, "@"); i > 0 {
+					name = r[:i]
+					req = r[i+1:]
+				}
+			}
+			deps = append(deps, resolve.RequirementVersion{
+				Type: typ,
+				VersionKey: resolve.VersionKey{
+					PackageKey: resolve.PackageKey{
+						System: resolve.NPM,
+						Name:   name,
+					},
+					VersionType: resolve.Requirement,
+					Version:     req,
+				},
+			})
+		}
+	}
+	addDeps(dependencies.Dependencies, dep.NewType())
+	addDeps(dependencies.DevDependencies, dep.NewType(dep.Dev))
+	addDeps(dependencies.OptionalDependencies, dep.NewType(dep.Opt))
+
+	peerType := dep.NewType()
+	peerType.AddAttr(dep.Scope, "peer")
+	addDeps(dependencies.PeerDependencies, peerType)
+
+	// The resolver expects bundleDependencies to be present as regular
+	// dependencies with a "*" version specifier, even if they were already
+	// in the regular dependencies.
+	bundleType := dep.NewType()
+	bundleType.AddAttr(dep.Scope, "bundle")
+	for _, name := range dependencies.BundleDependencies {
+		deps = append(deps, resolve.RequirementVersion{
+			Type: bundleType,
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.NPM,
+					Name:   name,
+				},
+				VersionType: resolve.Requirement,
+				Version:     "*",
+			},
+		})
+	}
+
+	// Correctly resolving the bundled dependencies would require downloading the package.
+	// Instead, call the fallback deps.dev client to get the bundled dependencies with mangled names.
+	if len(dependencies.BundleDependencies) > 0 {
+		fallbackReqs, err := c.fallback.Requirements(ctx, vk)
+		if err != nil {
+			// TODO: make some placeholder if the package doesn't exist on deps.dev
+			return nil, err
+		}
+		for _, req := range fallbackReqs {
+			if isNpmBundle(req.PackageKey) {
+				deps = append(deps, req)
+			}
+		}
+	}
+
+	resolve.SortDependencies(deps)
+
+	return deps, nil
+}
+
+func (c *NpmRegistryClient) MatchingVersions(ctx context.Context, vk resolve.VersionKey) ([]resolve.Version, error) {
+	if isNpmBundle(vk.PackageKey) { // bundled dependencies, fallback to deps.dev client
+		return c.fallback.MatchingVersions(ctx, vk)
+	}
+
+	versions, err := c.api.Versions(ctx, vk.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if concVer, ok := versions.Tags[vk.Version]; ok {
+		// matched a tag, return just the concrete version of the tag
+		return []resolve.Version{{
+			VersionKey: resolve.VersionKey{
+				PackageKey:  vk.PackageKey,
+				Version:     concVer,
+				VersionType: resolve.Concrete,
+			},
+		},
+		}, nil
```

Comment:
```suggestion
			},
		}}, nil
```
nit: the indents here are a bit confusing